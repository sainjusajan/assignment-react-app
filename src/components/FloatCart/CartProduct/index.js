import { useState } from 'react';
import crossIcon from './../../../assets/sprite_delete-icon.png'

const CartProduct = (props) => {

  const [isMouseOver, setIsMouseOver] = useState(false);
  

  const handleMouseOver = () => {
    setIsMouseOver(true);
  };

  const handleMouseOut = () => {
    setIsMouseOver(false);
  };

  
  const classes = ['shelf-item'];
  
  if (!!isMouseOver) {
    classes.push('shelf-item--mouseover');
  }
  
  return (
    <div className={classes.join(' ')}>
        <div
          style={{backgroundImage: `url(${crossIcon})`}}
          className="shelf-item__del"
          onMouseOver={() => handleMouseOver()}
          onMouseOut={() => handleMouseOut()}
          onClick={() => props.removeProduct(props.product)}
          />
    
        <div className="shelf-item__details">
          <p className="title">{props.product.name}</p>
          <p className="desc">
            {`${props.product.brand} | ${props.product.weight} Kg`} <br />
            Quantity: {props.product.quantity}
          </p>
        </div>
        <div className="shelf-item__price">
          <p>â‚¬ {props.product.price} </p>
          <div className='q-controls'>
            <button onClick={() => props.decrementCartItem(props.product)}>-</button>
            <button onClick={() => props.incrementCartItem(props.product)}>+</button>
          </div>
        </div>

      </div>
    );
  }

export default CartProduct;
