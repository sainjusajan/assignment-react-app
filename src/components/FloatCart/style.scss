.float-cart {
  background-color: #fff;
  box-sizing: border-box;
  z-index: 1;
  transition: right 0.2s;
  border-radius: 16px;
  .bag {
    width: 40px;
    height: 40px;
    position: relative;
    display: inline-block;
    vertical-align: middle;
    margin-right: 15px;
    background-repeat: no-repeat;
    background-size: contain;
    background-position: center;

    &--float-cart-closed {
      position: absolute;
      background-color: #fff;
      background-size: 50%;
      left: -60px;
      width: 60px;
      height: 60px;
      cursor: pointer;

      .bag__quantity {
        bottom: 5px;
        right: 10px;
      }
    }

    &__quantity {
      display: inline-block;
      width: 18px;
      height: 18px;
      color: #0c0b10;
      font-weight: bold;
      font-size: 0.7em;
      text-align: center;
      line-height: 18px;
      border-radius: 50%;
      background-color: #eabf00;
      position: absolute;
      bottom: -5px;
      right: 0px;
    }
  }

  &__header {
    color: #747474;
    box-sizing: border-box;
    text-align: center;
    padding: 45px 0;
    font-weight: 400;

    .header-title {
      font-size: 1.2em;
      vertical-align: middle;
    }
  }

  &__shelf-container {
    position: relative;
    min-height: 150px;
    padding-bottom: 200px;

    .shelf-empty {
      color: #747474;
      text-align: center;
      line-height: 40px;
    }

    .shelf-item {
      position: relative;
      box-sizing: border-box;
      padding: 5%;
      display: flex;
      justify-content: space-between;
      align-items: center;

      transition: background-color 0.2s, opacity 0.2s;

      &::before {
        content: "";
        width: 90%;
        height: 2px;
        background-color: rgba(0, 0, 0, 0.2);
        position: absolute;
        top: 0;
        left: 5%;
      }

      &--mouseover {
        background: #f3f3f3;

        .shelf-item__details {
          .title,
          .desc {
            text-decoration: line-through;
            opacity: 0.6;
          }
        }

        .shelf-item__price {
          text-decoration: line-through;
          opacity: 0.6;
        }
      }

      &__del {
        width: 16px;
        height: 16px;
        top: 15px;
        right: 5%;
        border-radius: 50%;
        position: absolute;
        background-size: auto 100%;
        background-repeat: no-repeat;
        z-index: 2;
        cursor: pointer;

        &:hover {
          background-position-x: -17px;
        }
      }

      &__details,
      &__price {
        display: inline-block;
        vertical-align: middle;
      }

      &__details {
        width: 57%;

        .title {
          color: #747474;
        }

        .desc {
          color: #5b5a5e;
          margin: 0;
        }
      }
      &__price {
        color: #eabf00;
        text-align: right;
      }
    }
  }

  &__footer {
    box-sizing: border-box;
    padding: 5%;
    height: 200px;
    z-index: 2;
    background-color: #ffffff;
    border-top: 1px solid #f1f1f1;
    border-bottom-left-radius: 16px;
    border-bottom-right-radius: 16px;

    &::before {
      content: "";
      width: 100%;
      height: 20px;
      display: block;
    }

    .sub,
    .sub-price {
      color: #5b5a5e;
      vertical-align: middle;
      display: inline-block;
    }

    .sub {
      width: 20%;
    }

    .sub-price {
      width: 80%;
      text-align: right;

      &__val,
      &__installment {
        margin: 0;
      }

      &__val {
        color: #eabf00;
        font-size: 22px;
      }
    }

    .buy-btn {
      color: #ececec;
      text-transform: uppercase;
      background-color: #009688;
      text-align: center;
      padding: 15px 0;
      margin-top: 40px;
      cursor: pointer;

      transition: background-color 0.2s;
      border-radius: 16px;

      &:hover {
        background-color: #009688;
      }
    }
  }
}

/* MAC scrollbar para desktop*/
@media screen and (min-width: 640px) {
  .float-cart__content::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 10px;
    background-color: rgba(0, 0, 0, 0.2);
    padding: 10px;
  }
  .float-cart__content::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: #0c0b10;
  }
}

.q-controls {
  display: flex;
  justify-content: center;
  align-items: center;
  button {
    height: 25px;
    width: 25px;
    background: #009688;
    color: white;
    border: 0;
    cursor: pointer;
    margin-left: 5px;
  }
}