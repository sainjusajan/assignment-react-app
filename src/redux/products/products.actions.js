import ProductsActionTypes from "./products.types"

export const getProductsListStart = () => ({
  type: ProductsActionTypes.GET_PRODUCTS_LIST_START,
})


export const getProductsListSuccess = (payload) => ({
  type: ProductsActionTypes.GET_PRODUCTS_LIST_SUCCESS,
  payload
})

export const getProductsListFailure = (error) => ({
  type: ProductsActionTypes.GET_PRODUCTS_LIST_FAILURE,
  payload: error
})


