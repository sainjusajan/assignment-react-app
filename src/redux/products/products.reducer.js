import ProductsActionTypes from "./products.types";

const INITIAL_STATE = {
  items: false,
  isItemsLoading: false,
  itemsErrorMessage: "",
};

const productsListReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ProductsActionTypes.GET_PRODUCTS_LIST_START:
      return {
        ...state,
        isItemsLoading: true,
        itemsErrorMessage: "",
      };
      
    case ProductsActionTypes.GET_PRODUCTS_LIST_SUCCESS:
      return {
        ...state,
        isItemsLoading: false,
        items: action.payload.items,
      };
      
    case ProductsActionTypes.GET_PRODUCTS_LIST_FAILURE:
      return {
        ...state,
        isItemsLoading: false,
        itemsErrorMessage: action.payload,
      };

    default:
      return state
  }
};


export default productsListReducer