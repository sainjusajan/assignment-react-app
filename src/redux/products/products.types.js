const ProductsActionTypes = {
  GET_PRODUCTS_LIST_START : 'App/Products/GET_PRODUCTS_LIST_START',
  GET_PRODUCTS_LIST_SUCCESS : 'App/Products/GET_PRODUCTS_LIST_SUCCESS',
  GET_PRODUCTS_LIST_FAILURE : 'App/Products/GET_PRODUCTS_LIST_FAILURE',
}

export default ProductsActionTypes