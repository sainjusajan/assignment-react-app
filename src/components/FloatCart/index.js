import cartIcon from '../../assets/Background.png'
import { connect } from "react-redux";
import {
  selectCartItems,
  selectCartItemsCount,
  selectCartTotal,
} from "../../redux/Cart/cart.selectors";

import { createStructuredSelector } from "reselect";
import CartProduct from "./CartProduct";

import "./style.scss";
import { addItem, clearItemFromCart, removeItem } from "../../redux/Cart/cart.actions";

const FloatCart = ({
  cart,
  cartItemsCount,
  cartItemsTotal,
  onRemoveItemFromCart,
  onIncrementCartItem,
  onDecrementCartItem
}) => {

  const removeProduct = (product) => {
    onRemoveItemFromCart(product)
  };
  const decrementCartItem = (product) => {
    onDecrementCartItem(product)
  };
  const incrementCartItem = (product) => {
    const itm = {...product, quantity: 1};
    onIncrementCartItem(itm);
  };

  const proceedToCheckout = () => {
    alert('Going to checkout!!')
  };

  let classes = ["float-cart"];

  return (
    <div className={classes.join(" ")}>
  

      <div className="float-cart__content">
        <div className="float-cart__header">
          <span className="bag" style={{backgroundImage: `url(${cartIcon})`}}>
            <span className="bag__quantity">{cartItemsCount}</span>
          </span>
          <span className="header-title">Bag</span>
        </div>

        <div className="float-cart__shelf-container">
          {cart.map((p) => {
            return (
              <CartProduct
                product={p}
                removeProduct={removeProduct}
                decrementCartItem={decrementCartItem}
                incrementCartItem={incrementCartItem}
                key={p.id}
              />
            );
          })}
          {!cart.length && (
            <p className="shelf-empty">
              Add some products in the bag <br />
            </p>
          )}
        </div>

        <div className="float-cart__footer">
          <div className="sub">TOTAL</div>
          <div className="sub-price">
            <p className="sub-price__val">â‚¬ {cartItemsTotal}</p>
          </div>
          <div onClick={() => proceedToCheckout()} className="buy-btn">
            Checkout
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  cart: selectCartItems,
  cartItemsTotal: selectCartTotal,
  cartItemsCount: selectCartItemsCount,
});

const mapDispatchToProps = (dispatch) => ({
  onRemoveItemFromCart: (product) => dispatch(clearItemFromCart(product)),
  onDecrementCartItem: (product) => dispatch(removeItem(product)),
  onIncrementCartItem: (product) => dispatch(addItem(product))
});

export default connect(mapStateToProps, mapDispatchToProps)(FloatCart);
