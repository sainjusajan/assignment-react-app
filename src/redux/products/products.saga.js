import { takeLatest, put, all, call } from 'redux-saga/effects'
import request from '../../utils/request';
import { getProductsListSuccess } from './products.actions';
import ProductsActionTypes from './products.types'


export function* getProducts() {
  try {
    const data = yield call(request, "../../data/products.json", {
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
    });
    yield put(getProductsListSuccess(data))
  } catch (error) {
    console.log(error);
  }
}

export function* onFetchProductStart() {
  yield takeLatest(ProductsActionTypes.GET_PRODUCTS_LIST_START, getProducts)
}


export function* productsSaga() {
  yield all([
    call(onFetchProductStart),
  ])
}
