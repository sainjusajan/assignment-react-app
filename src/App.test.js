import React from "react";
import { Provider } from "react-redux";
import "@testing-library/jest-dom";

import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import "./../setupTest";
import * as actions from "./redux/products/products.actions";
import reducer from "./redux/products/products.reducer";
import ProductsActionTypes from "./redux/products/products.types";
import FloatCart from "./components/FloatCart";
import ProductDetailPage from "./pages/ProductDetail";
import CheckoutPage from "./pages/Checkout";

describe("ACTIONS", () => {
  it("should create an action with correct type", () => {
    const expectedAction = {
      type: "App/Products/GET_PRODUCTS_LIST_START",
    };
    expect(actions.getProductsListStart()).toEqual(expectedAction);
  });
});

describe("REDUCER", () => {
  it("should return the initial state", () => {
    expect(
      reducer(undefined, {
        items: false,
        isItemsLoading: false,
        itemsErrorMessage: "",
      })
    ).toEqual({
      items: false,
      isItemsLoading: false,
      itemsErrorMessage: "",
    });
  });
  it('should handle "GET_PRODUCTS_LIST_START" action', () => {
    expect(
      reducer({}, { type: ProductsActionTypes.GET_PRODUCTS_LIST_START })
    ).toHaveProperty("isItemsLoading");
  });
});

describe("RENDER", () => {
  let mockStore;
  let wrapper;
  let store;
  beforeEach(() => {
    mockStore = configureStore();
    store = mockStore({
      cart: {
        cartItems: [
          {
            id: 1,
            name: "Philips hue bulb",
            brand: "Philips",
            price: "500",
            available: true,
            weight: 0.2,
            options: {
              color: "white",
              power: 6.5,
            },
            quantity: 2,
          },
        ],
        isAdded: false,
        isRemoved: false,
      },
    });

    wrapper = mount(
      <Provider store={store}>
        <FloatCart />
      </Provider>
    );
  });

  it("should increment product on checkout page", () => {
    const button = wrapper.findWhere((node) => {
      return node.type() === "button" && node.text() === "+";
    });
    expect(button).toHaveLength(1);
    button.simulate("click");
    const delay = (time) => new Promise((resolve) => setTimeout(resolve, time));
    delay(5000).then(() => {
      expect(wrapper.find(".bag__quantity").text()).toEqual("3");
    });
  });
  it("should decrement product on checkout page", () => {
    const button = wrapper.findWhere((node) => {
      return node.type() === "button" && node.text() === "-";
    });
    expect(button).toHaveLength(1);
    button.simulate("click");
    const delay = (time) => new Promise((resolve) => setTimeout(resolve, time));
    delay(5000).then(() => {
      expect(wrapper.find(".bag__quantity").text()).toEqual("1");
    });
  });

  it("should render correctly checkout page", () => {
    const mockStore = configureStore();
    const store = mockStore({
      cart: {
        cartItems: [],
        isAdded: false,
        isRemoved: false,
      },
    });
    const wrapper = mount(
      <Provider store={store}>
        <CheckoutPage />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
