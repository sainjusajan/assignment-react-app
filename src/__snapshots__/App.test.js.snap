// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDER should render correctly checkout page 1`] = `ReactWrapper {}`;
